govuk-one-login/authentication-frontend
# BAU: Bump redis dependencies

## What?

Updated [redis](https://npmjs.org/package/redis) from `^3.1.2` to `^4.6.13` and [connect-redis](https://npmjs.org/package/connect-redis) from `^6.0.0` to `^7.1.1`.

There are two main changes here:
- Both packages introduced their own Typescript typings, thus the removal of `@types/redis` and `@types/connect-redis`.
- All functions of `redis` return promises by default. In v3, calling `createClient()` would automatically attempt a connection. In v4, we need to call `client.connect()`, which returns a Promise.
  - This update implementation does not attempt to wait until the connection promise to be fulfilled or rejected before moving on, thus keep the same behaviour pre-update. Connection errors are handled elsewhere when redis is queried (in this case within the `connect-redis` package). 

Also removed unused properties from the `RedisConfig` type and a minor restructure of how `RedisConfig` object is created, centralising it in the single `getRedisConfig()` function.

Note there seems to be a difference in the error thrown when the redis service is halted, however both result in a Node service crash.

The following test scenario was performed for both versions:
- Open http://localhost:2000
- See the redirect
- Kill the `redis-1` Docker container
- Node service crashes with the following errors:

For v3:
```
[Node] Error: connect ECONNREFUSED 127.0.0.1:6389
[Node]     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1300:16)
[Node] Emitted 'error' event on RedisClient instance at:
[Node]     at RedisClient.on_error (/Users/matt.votsikas-mclean/Projects/authentication-frontend/node_modules/redis/index.js:342:14)
[Node]     at Socket.<anonymous> (/Users/matt.votsikas-mclean/Projects/authentication-frontend/node_modules/redis/index.js:223:14)
[Node]     at Socket.emit (node:events:513:28)
[Node]     at Socket.emit (node:domain:489:12)
[Node]     at emitErrorNT (node:internal/streams/destroy:151:8)
[Node]     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
[Node]     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
[Node]   errno: -61,
[Node]   code: 'ECONNREFUSED',
[Node]   syscall: 'connect',
[Node]   address: '127.0.0.1',
[Node]   port: 6389
[Node] }
[Node] 
[Node] Node.js v18.12.1
[Node] [nodemon] app crashed - waiting for file changes before starting...
```

For v4:
```
[Node] SocketClosedUnexpectedlyError: Socket closed unexpectedly
[Node]     at Socket.<anonymous> (/Users/matt.votsikas-mclean/Projects/authentication-frontend/node_modules/@redis/client/dist/lib/client/socket.js:194:118)
[Node]     at Object.onceWrapper (node:events:628:26)
[Node]     at Socket.emit (node:events:513:28)
[Node]     at Socket.emit (node:domain:489:12)
[Node]     at TCP.<anonymous> (node:net:313:12)
[Node] Emitted 'error' event on Commander instance at:
[Node]     at RedisSocket.<anonymous> (/Users/matt.votsikas-mclean/Projects/authentication-frontend/node_modules/@redis/client/dist/lib/client/index.js:412:14)
[Node]     at RedisSocket.emit (node:events:513:28)
[Node]     at RedisSocket.emit (node:domain:489:12)
[Node]     at RedisSocket._RedisSocket_onSocketError (/Users/matt.votsikas-mclean/Projects/authentication-frontend/node_modules/@redis/client/dist/lib/client/socket.js:218:10)
[Node]     at Socket.<anonymous> (/Users/matt.votsikas-mclean/Projects/authentication-frontend/node_modules/@redis/client/dist/lib/client/socket.js:194:107)
[Node]     at Object.onceWrapper (node:events:628:26)
[Node]     [... lines matching original stack trace ...]
[Node]     at TCP.<anonymous> (node:net:313:12)
[Node] 
[Node] Node.js v18.12.1
[Node] [nodemon] app crashed - waiting for file changes before starting...
```

## Why?

We should endeavour to keep up-to-date with dependencies, especially key ones like our connection to redis.

## Performance Analysis have been informed of the change

- [x] Performance Analysis have been informed of the change
